File: kickstart/2019H/a.js
const readline = require('readline');
process.stdin.setEncoding('utf-8');
const rl = readline.createInterface({input: process.stdin});
let lineNumber = 0;
let caseNumber = 0;
rl.on('line', (line) => {
    lineNumber++;
    if (lineNumber >= 3 && lineNumber % 2 === 1) {
        caseNumber++;
        const map = {};
        let hIndex = 1;
        let papersAboveHIndex = 0;
        let result = line
            .split(' ')
            .map(value => {
                if (value > hIndex) {
                    map[value] = map[value] ? map[value] + 1 : 1;
                    papersAboveHIndex += 1;
                    while (papersAboveHIndex > hIndex) {
                        hIndex += 1;
                        papersAboveHIndex -= map[hIndex] ? map[hIndex] : 0;
                    }
                }
                return hIndex;
            });
        console.log('Case #' + caseNumber + ': ' + result.join(' '));
    }
});

File: kickstart/2019H/a.py
import heapq 
for t in range(int(input())):
    n = int(input())
    A = [int(x) for x in input().split()]    
    C = list([A[0]])
    k = 1
    R = [1]
    for i in range(1,n):        
        if (A[i] >= k):
            heapq.heappush(C,A[i])
            p = heapq.heappop(C)
            if (p >= k+1):
                heapq.heappush(C, p)
                k+=1
        R.append(k)        
    print ("Case #%d: %s" % (t+1, ' '.join([str(x) for x in R])))